### PR TITLE
Vn 236 raise event for parent when children value is changed

### PR DESCRIPTION
**Pull request by votuanphong1983**
Raise event for parent path if children value  is changed. E.g...

> settings.watch('foo')

 => will be populated if user set settings.set('foo.bar', newValue)

### CODE REVIEW

File: lib/settings.js [hunk 1]
```diff
@@ -94,12 +94,30 @@ class Settings extends EventEmitter {
...
   }
 
+  _buildWatchPaths(path) {
+    const children = path.split('.');
+    const paths = [];
+
+    for (let i = 0; i < children.length; i++) {
+      const path = _.take(children, i + 1);
+      paths.push(_.join(path, '.'));
+    }
```
Comment (pandafox) [inline lib/settings.js]: I think the notification should be more specific to less specific

ex:
a.b.c.d
a.b.c
a.b
a

Comment (pandafox) [reply lib/settings.js]: Ugh, instead of reverse, just change the for loop order from children.length to 0?

Also, perhaps rename children to parts.

**Review by pandafox — CHANGES_REQUESTED** [review]

File: lib/settings.js [hunk 2]
```diff
@@ -94,12 +94,30 @@ class Settings extends EventEmitter {
...
     const settingsFilePath = this._getSettingsFilePath();
     jsonfile.writeFileSync(settingsFilePath, this._cacheObj);
-    const relatedHandler = keyPath ?
-      _.filter(this._watchHandlers, f => f.keyPath.startsWith(keyPath)) :
-      this._watchHandlers;
+    let relatedHandler = [];
+    if (keyPath) {
+      const paths = this._buildWatchPaths(keyPath);
+      relatedHandler =
+      _.filter(this._watchHandlers, f => _.some(paths, c => f.keyPath.startsWith(c)));
```
Comment (pandafox) [inline lib/settings.js]: I don't get why we need to do the path splitting, can you give a specific example of how your change catches and the old one doesn't?

The test case that you have seems to suggest that it'll still work for old code? (ex: without splitting)

Comment (votuanphong1983) [reply lib/settings.js]:
```
Setting = {
   config : {
      isLogToFile : true,
      mining: {
          isAutoMining: false
      }
   }
}
```
setting.watch('config')
setting.set('config.isLogToFile', true)

`Old setting`: watch does not received change event
`New setting`: watch receives change event

That mean: new setting can raise to parent's watcher if there is any change in child field.


Our old setting can only raise event in below case 

watch('config.mining.isAutoMining') => received change if value is changed
watch ('config.isLogToFile) => received change if value is changed
 
setting.set('config', {xxxx})

Comment (pandafox) [reply lib/settings.js]: Okay, that's not what's included in the testing though.

Also, if this is what we are testing, shouldn't the following work as well?
```
actual path.startWith(watch key) || watch key.startWith(actual path)
```

Comment (votuanphong1983) [reply lib/settings.js]: ah...i think it will work as well.

**Review by pandafox — APPROVED** [review]

**Pull request merged**